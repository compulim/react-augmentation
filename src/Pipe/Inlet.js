import PropTypes from 'prop-types';
import React     from 'react';

export default class Inlet extends React.PureComponent {
  constructor(props, context) {
    super(props, context);

    this.state = {
      name: props.name
    };
  }

  componentDidMount() {
    this.context.pipe.update(this.props.name, this.props.children);
  }

  componentWillReceiveProps(nextProps) {
    const { name: nextName } = nextProps;

    if (nextName !== this.state.name) {
      this.context.pipe.update(this.state.name, null);

      this.setState(() => ({ name: nextName }), () => {
        this.context.pipe.update(nextName, nextProps.children);
      });
    } else if (nextProps.children !== this.props.children) {
      this.context.pipe.update(nextName, nextProps.children);
    }
  }

  componentWillUnmount() {
    this.context.pipe.update(this.state.name, null);
  }

  render() {
    return false;
  }
}

Inlet.contextTypes = {
  pipe: PropTypes.shape({
    update: PropTypes.func
  })
};

Inlet.propTypes = {
  name: PropTypes.string
};
