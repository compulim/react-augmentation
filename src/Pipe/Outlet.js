import PropTypes from 'prop-types';
import React     from 'react';

export default class Outlet extends React.PureComponent {
  constructor(props, context) {
    super(props, context);

    this.handleUpdate = this.handleUpdate.bind(this);

    this.state = {
      content: null
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.name !== this.props.name) {
      this.unsubscribe(this.props);
      this.subscribe(nextProps);
    }
  }

  componentWillMount() {
    this.props.name && this.subscribe(this.props);
  }

  componentWillUnmount() {
    this.unsubscribe(this.props);
  }

  subscribe(props) {
    const nextContent = this.context.pipe.subscribe(props.name, this.handleUpdate);

    this.setState(() => ({ content: nextContent }));
  }

  unsubscribe(props) {
    this.context.pipe.unsubscribe(props.name, this.handleUpdate);
  }

  handleUpdate(nextContent) {
    this.setState(() => ({ content: nextContent }));
  }

  render() {
    const { name, ...otherProps } = this.props;

    return (
      <div { ...otherProps }>
        { this.state.content || this.props.children }
      </div>
    );
  }
}

Outlet.contextTypes = {
  pipe: PropTypes.shape({
    subscribe  : PropTypes.func,
    unsubscribe: PropTypes.func
  })
};

Outlet.propTypes = {
  name: PropTypes.string
};
