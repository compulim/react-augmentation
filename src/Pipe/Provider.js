import PropTypes from 'prop-types';
import React     from 'react';

export default class PipeProvider extends React.PureComponent {
  constructor(props, context) {
    super(props, context);

    this.contents = {};
    this.outlets = {};
  }

  getChildContext() {
    return {
      pipe: {
        subscribe: (name, observer) => {
          (this.outlets[name] || (this.outlets[name] = [])).push(observer);

          return this.contents[name];
        },
        unsubscribe: (name, observer) => {
          const targetOutlets = this.outlets[name];

          if (targetOutlets) {
            this.outlets[name] = targetOutlets.filter(o => o !== observer);

            if (!this.outlets[name].length) {
              delete this.outlets[name];
            }
          }
        },
        update: (name, children) => {
          if (children) {
            this.contents[name] = children;
          } else {
            delete this.contents[name];
          }

          if (this.outlets[name]) {
            this.outlets[name].forEach(observer => observer(children));
          }
        }
      }
    };
  }

  render() {
    return (
      <div { ...this.props }>
        { this.props.children }
      </div>
    );
  }
};

PipeProvider.childContextTypes = {
  pipe: PropTypes.shape({
    subscribe  : PropTypes.func,
    unsubscribe: PropTypes.func,
    update     : PropTypes.func
  })
};
