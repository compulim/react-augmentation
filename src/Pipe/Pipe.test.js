import React    from 'react';
import renderer from 'react-test-renderer';

import Inlet    from './Inlet';
import Outlet   from './Outlet';
import Provider from './Provider';

it('pipes correctly', () => {
  const tree = renderer.create(
    <Provider>
      <div className="inlet">
        <Inlet name="aloha" className="not-passed">
          <p>How are you?</p>
          <p>I am fine.</p>
        </Inlet>
      </div>
      <div className="outlet">
        <Outlet name="aloha" className="passed" />
      </div>
    </Provider>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('switch pipe correctly', () => {
  class TestHost extends React.Component {
    render() {
      return (
        <Provider>
          <h1>{ this.props.pipeName }</h1>
          <div className="inlet">
            <Inlet name={ this.props.pipeName }>
              <p>How are you?</p>
              <p>I am fine.</p>
            </Inlet>
          </div>
          <div className="first-outlet">
            <Outlet name="first" />
          </div>
          <div className="second-outlet">
            <Outlet name="second" />
          </div>
        </Provider>
      );
    }
  }

  const component = renderer.create(
    <TestHost pipeName="first" />
  );

  expect(component.toJSON()).toMatchSnapshot();

  component.update(
    <TestHost pipeName="second" />
  );

  expect(component.toJSON()).toMatchSnapshot();
});

it('cross switch pipe correctly', () => {
  class TestHost extends React.Component {
    render() {
      return (
        <Provider>
          <h1>{ this.props.firstPipeName } x { this.props.secondPipeName }</h1>
          <div className="first-inlet">
            <Inlet name={ this.props.firstPipeName }>
              <p>How are you?</p>
            </Inlet>
          </div>
          <div className="second-inlet">
            <Inlet name={ this.props.secondPipeName }>
              <p>I am fine.</p>
            </Inlet>
          </div>
          <div className="first-outlet">
            <Outlet name="first" />
          </div>
          <div className="second-outlet">
            <Outlet name="second" />
          </div>
        </Provider>
      );
    }
  }

  const component = renderer.create(
    <TestHost firstPipeName="first" secondPipeName="second" />
  );

  expect(component.toJSON()).toMatchSnapshot();

  component.update(
    <TestHost firstPipeName="second" secondPipeName="first" />
  );

  expect(component.toJSON()).toMatchSnapshot();
});

it('multiplex correctly', () => {
  const tree = renderer.create(
    <Provider>
      <div className="inlet">
        <Inlet name="aloha">
          <p>How are you?</p>
          <p>I am fine.</p>
        </Inlet>
      </div>
      <div className="first-outlet">
        <Outlet name="aloha" />
      </div>
      <div className="second-outlet">
        <Outlet name="aloha" />
      </div>
    </Provider>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('propagate content change correctly', () => {
  const component = renderer.create(
    <Provider>
      <div className="inlet">
        <Inlet name="greeting">
          <p>Hello, World!</p>
        </Inlet>
      </div>
      <div className="outlet">
        <Outlet name="greeting" />
      </div>
    </Provider>
  );

  expect(component.toJSON()).toMatchSnapshot();

  component.update(
    <Provider>
      <div className="inlet">
        <Inlet name="greeting">
          <p>Aloha!</p>
        </Inlet>
      </div>
      <div className="outlet">
        <Outlet name="greeting" />
      </div>
    </Provider>
  );

  expect(component.toJSON()).toMatchSnapshot();
});

it('inlet join and leave pipe correctly', () => {
  const component = renderer.create(
    <Provider>
      <div className="inlet">
        <Inlet>
          <p>Hello, World!</p>
        </Inlet>
      </div>
      <div className="outlet">
        <Outlet name="greeting" />
      </div>
    </Provider>
  );

  expect(component.toJSON()).toMatchSnapshot();

  component.update(
    <Provider>
      <div className="inlet">
        <Inlet name="greeting">
          <p>Hello, World!</p>
        </Inlet>
      </div>
      <div className="outlet">
        <Outlet name="greeting" />
      </div>
    </Provider>
  );

  expect(component.toJSON()).toMatchSnapshot();

  component.update(
    <Provider>
      <div className="inlet">
        <Inlet>
          <p>Hello, World!</p>
        </Inlet>
      </div>
      <div className="outlet">
        <Outlet name="greeting" />
      </div>
    </Provider>
  );

  expect(component.toJSON()).toMatchSnapshot();
});

it('outlet join and leave pipe correctly', () => {
  const component = renderer.create(
    <Provider>
      <div className="inlet">
        <Inlet name="greeting">
          <p>Hello, World!</p>
        </Inlet>
      </div>
      <div className="outlet">
        <Outlet />
      </div>
    </Provider>
  );

  expect(component.toJSON()).toMatchSnapshot();

  component.update(
    <Provider>
      <div className="inlet">
        <Inlet name="greeting">
          <p>Hello, World!</p>
        </Inlet>
      </div>
      <div className="outlet">
        <Outlet name="greeting" />
      </div>
    </Provider>
  );

  expect(component.toJSON()).toMatchSnapshot();

  component.update(
    <Provider>
      <div className="inlet">
        <Inlet name="greeting">
          <p>Hello, World!</p>
        </Inlet>
      </div>
      <div className="outlet">
        <Outlet />
      </div>
    </Provider>
  );

  expect(component.toJSON()).toMatchSnapshot();
});

it('should unmount pipe correctly', () => {
  const component = renderer.create(
    <Provider>
      <div className="inlet">
        <Inlet name="greeting">
          <p>Hello, World!</p>
        </Inlet>
      </div>
      <div className="outlet">
        <Outlet name="greeting" />
      </div>
    </Provider>
  );

  expect(component.toJSON()).toMatchSnapshot();

  component.update(
    <Provider>
      <div className="inlet">
      </div>
      <div className="outlet">
        <Outlet name="greeting" />
      </div>
    </Provider>
  );

  expect(component.toJSON()).toMatchSnapshot();

  component.update(
    <Provider>
      <div className="inlet">
      </div>
      <div className="outlet">
      </div>
    </Provider>
  );

  expect(component.toJSON()).toMatchSnapshot();
});

it('shows outlet children if nothing from inlet', () => {
  const component = renderer.create(
    <Provider>
      <div className="inlet">
        <Inlet name="aloha" />
      </div>
      <div className="outlet">
        <Outlet name="aloha">
          <p>Default content</p>
        </Outlet>
      </div>
    </Provider>
  );

  expect(component.toJSON()).toMatchSnapshot();

  component.update(
    <Provider>
      <div className="inlet">
        <Inlet name="aloha">
          <p>Overriden content</p>
        </Inlet>
      </div>
      <div className="outlet">
        <Outlet name="aloha">
          <p>Default content</p>
        </Outlet>
      </div>
    </Provider>
  );

  expect(component.toJSON()).toMatchSnapshot();
});
