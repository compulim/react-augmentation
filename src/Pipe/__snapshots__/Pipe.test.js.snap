// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cross switch pipe correctly 1`] = `
<div>
  <h1>
    first
     x 
    second
  </h1>
  <div
    className="first-inlet"
  />
  <div
    className="second-inlet"
  />
  <div
    className="first-outlet"
  >
    <div>
      <p>
        How are you?
      </p>
    </div>
  </div>
  <div
    className="second-outlet"
  >
    <div>
      <p>
        I am fine.
      </p>
    </div>
  </div>
</div>
`;

exports[`cross switch pipe correctly 2`] = `
<div>
  <h1>
    second
     x 
    first
  </h1>
  <div
    className="first-inlet"
  />
  <div
    className="second-inlet"
  />
  <div
    className="first-outlet"
  >
    <div>
      <p>
        I am fine.
      </p>
    </div>
  </div>
  <div
    className="second-outlet"
  >
    <div>
      <p>
        How are you?
      </p>
    </div>
  </div>
</div>
`;

exports[`inlet join and leave pipe correctly 1`] = `
<div>
  <div
    className="inlet"
  />
  <div
    className="outlet"
  >
    <div />
  </div>
</div>
`;

exports[`inlet join and leave pipe correctly 2`] = `
<div>
  <div
    className="inlet"
  />
  <div
    className="outlet"
  >
    <div>
      <p>
        Hello, World!
      </p>
    </div>
  </div>
</div>
`;

exports[`inlet join and leave pipe correctly 3`] = `
<div>
  <div
    className="inlet"
  />
  <div
    className="outlet"
  >
    <div />
  </div>
</div>
`;

exports[`multiplex correctly 1`] = `
<div>
  <div
    className="inlet"
  />
  <div
    className="first-outlet"
  >
    <div>
      <p>
        How are you?
      </p>
      <p>
        I am fine.
      </p>
    </div>
  </div>
  <div
    className="second-outlet"
  >
    <div>
      <p>
        How are you?
      </p>
      <p>
        I am fine.
      </p>
    </div>
  </div>
</div>
`;

exports[`outlet join and leave pipe correctly 1`] = `
<div>
  <div
    className="inlet"
  />
  <div
    className="outlet"
  >
    <div />
  </div>
</div>
`;

exports[`outlet join and leave pipe correctly 2`] = `
<div>
  <div
    className="inlet"
  />
  <div
    className="outlet"
  >
    <div>
      <p>
        Hello, World!
      </p>
    </div>
  </div>
</div>
`;

exports[`outlet join and leave pipe correctly 3`] = `
<div>
  <div
    className="inlet"
  />
  <div
    className="outlet"
  >
    <div />
  </div>
</div>
`;

exports[`pipes correctly 1`] = `
<div>
  <div
    className="inlet"
  />
  <div
    className="outlet"
  >
    <div
      className="passed"
    >
      <p>
        How are you?
      </p>
      <p>
        I am fine.
      </p>
    </div>
  </div>
</div>
`;

exports[`propagate content change correctly 1`] = `
<div>
  <div
    className="inlet"
  />
  <div
    className="outlet"
  >
    <div>
      <p>
        Hello, World!
      </p>
    </div>
  </div>
</div>
`;

exports[`propagate content change correctly 2`] = `
<div>
  <div
    className="inlet"
  />
  <div
    className="outlet"
  >
    <div>
      <p>
        Aloha!
      </p>
    </div>
  </div>
</div>
`;

exports[`should unmount pipe correctly 1`] = `
<div>
  <div
    className="inlet"
  />
  <div
    className="outlet"
  >
    <div>
      <p>
        Hello, World!
      </p>
    </div>
  </div>
</div>
`;

exports[`should unmount pipe correctly 2`] = `
<div>
  <div
    className="inlet"
  />
  <div
    className="outlet"
  >
    <div />
  </div>
</div>
`;

exports[`should unmount pipe correctly 3`] = `
<div>
  <div
    className="inlet"
  />
  <div
    className="outlet"
  />
</div>
`;

exports[`shows outlet children if nothing from inlet 1`] = `
<div>
  <div
    className="inlet"
  />
  <div
    className="outlet"
  >
    <div>
      <p>
        Default content
      </p>
    </div>
  </div>
</div>
`;

exports[`shows outlet children if nothing from inlet 2`] = `
<div>
  <div
    className="inlet"
  />
  <div
    className="outlet"
  >
    <div>
      <p>
        Overriden content
      </p>
    </div>
  </div>
</div>
`;

exports[`switch pipe correctly 1`] = `
<div>
  <h1>
    first
  </h1>
  <div
    className="inlet"
  />
  <div
    className="first-outlet"
  >
    <div>
      <p>
        How are you?
      </p>
      <p>
        I am fine.
      </p>
    </div>
  </div>
  <div
    className="second-outlet"
  >
    <div />
  </div>
</div>
`;

exports[`switch pipe correctly 2`] = `
<div>
  <h1>
    second
  </h1>
  <div
    className="inlet"
  />
  <div
    className="first-outlet"
  >
    <div />
  </div>
  <div
    className="second-outlet"
  >
    <div>
      <p>
        How are you?
      </p>
      <p>
        I am fine.
      </p>
    </div>
  </div>
</div>
`;
